File: cpp/1609_is_even_odd_tree/is_even_odd_tree.cpp
/*
 * @Date: 2021-12-25 01:39:02
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-25 06:38:43
 */

#include <cassert>
#include <climits>
#include <queue>

using namespace std;

//  Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  bool isEvenOddTree(TreeNode *root) {
    queue<TreeNode *> qu;
    qu.push(root);
    int level = 0;
    while (!qu.empty()) {
      int size = qu.size();
      int prev = level % 2 == 0 ? INT_MIN : INT_MAX;
      for (int i = 0; i < size; i++) {
        TreeNode *node = qu.front();
        qu.pop();
        int value = node->val;
        if (level % 2 == value % 2 || (level % 2 == 0 && value <= prev) ||
            (level % 2 == 1 && value >= prev)) {
          return false;
        }

        prev = value;
        if (node->left != nullptr) {
          qu.push(node->left);
        }
        if (node->right != nullptr) {
          qu.push(node->right);
        }
      }
      level++;
    }
    return true;
  }
};

int main() {
  {
    TreeNode *root = new TreeNode(
        1,
        new TreeNode(10, new TreeNode(3, new TreeNode(12), new TreeNode(8)),
                     nullptr),
        new TreeNode(4, new TreeNode(7, new TreeNode(6), nullptr),
                     new TreeNode(9, nullptr, new TreeNode(2))));
    assert(Solution().isEvenOddTree(root) == true);
  }
  {
    TreeNode *root =
        new TreeNode(5, new TreeNode(4, new TreeNode(3), new TreeNode(3)),
                     new TreeNode(2, new TreeNode(7), nullptr));
    assert(Solution().isEvenOddTree(root) == false);
  }
  {
    TreeNode *root =
        new TreeNode(5, new TreeNode(9, new TreeNode(3), new TreeNode(5)),
                     new TreeNode(1, new TreeNode(7), nullptr));
    assert(Solution().isEvenOddTree(root) == false);
  }
  {
    TreeNode *root = new TreeNode(1);
    assert(Solution().isEvenOddTree(root) == true);
  }
  {
    TreeNode *root = new TreeNode(
        11,
        new TreeNode(
            8,
            new TreeNode(1, new TreeNode(30, new TreeNode(17), nullptr),
                         new TreeNode(20)),
            new TreeNode(3, new TreeNode(18), new TreeNode(16))),
        new TreeNode(6, new TreeNode(9, new TreeNode(12), new TreeNode(10)),
                     new TreeNode(11, new TreeNode(4), new TreeNode(2))));
    assert(Solution().isEvenOddTree(root) == true);
  }
}

File: rust/1609_is_even_odd_tree/is_even_odd_tree.rs
/*
 * @Date: 2021-12-25 01:39:07
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-25 06:35:23
 */

use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;

impl Solution {
    pub fn is_even_odd_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        use std::collections::VecDeque;
        let mut queue = VecDeque::new();
        if let Some(root) = root {
            queue.push_back(root.clone());
        };
        let mut level = 0;
        while !queue.is_empty() {
            let len = queue.len();
            let mut prev = if level % 2 == 0 { i32::MIN } else { i32::MAX };
            for _ in 0..len {
                if let Some(node) = queue.pop_front() {
                    let node = node.borrow();
                    let value = node.val;
                    if value % 2 == level % 2
                        || level % 2 == 0 && value <= prev
                        || level % 2 == 1 && value >= prev
                    {
                        return false;
                    }

                    prev = value;
                    if let Some(left) = &node.left {
                        queue.push_back(left.clone());
                    }
                    if let Some(right) = &node.right {
                        queue.push_back(right.clone());
                    }
                }
            }
            level += 1;
        }
        true
    }
}

fn main() {
    {
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 1,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 10,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 12,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 8,
                        left: None,
                        right: None,
                    }))),
                }))),
                right: None,
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode {
                val: 4,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 7,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 6,
                        left: None,
                        right: None,
                    }))),
                    right: None,
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 9,
                    left: None,
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 2,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
        })));

        assert_eq!(Solution::is_even_odd_tree(root), true);
    }

    {
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 5,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 4,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: None,
                    right: None,
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: None,
                    right: None,
                }))),
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode {
                val: 2,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 7,
                    left: None,
                    right: None,
                }))),
                right: None,
            }))),
        })));
        assert_eq!(Solution::is_even_odd_tree(root), false);
    }

    {
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 5,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 9,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: None,
                    right: None,
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 5,
                    left: None,
                    right: None,
                }))),
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode {
                val: 1,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 7,
                    left: None,
                    right: None,
                }))),
                right: None,
            }))),
        })));
        assert_eq!(Solution::is_even_odd_tree(root), false);
    }

    {
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 1,
            left: None,
            right: None,
        })));
        assert_eq!(Solution::is_even_odd_tree(root), true);
    }

    {
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 11,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 8,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 30,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 20,
                        left: None,
                        right: None,
                    }))),
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 18,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 16,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode {
                val: 6,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 9,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 12,
                        left: Some(Rc::new(RefCell::new(TreeNode {
                            val: 17,
                            left: None,
                            right: None,
                        }))),
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 10,
                        left: None,
                        right: None,
                    }))),
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 11,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 4,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 2,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
        })));
        assert_eq!(Solution::is_even_odd_tree(root), true);
    }
}
